plugins/auth-backend-module-pinniped-provider/src/authenticator.ts

```diff
@@ -0,0 +1,167 @@
+    const state = { ...decodedState, audience: stringifiedAudience };
+    const options: Record<string, string> = {
+      scope:
+        input.scope ||
+        'openid pinniped:request-audience username offline_access',
+      state: encodeOAuthState(state),
+    };
+
+    return new Promise((resolve, reject) => {
+      strategy.redirect = (url: string) => {
+        resolve({ url });
+      };
+      strategy.error = (error: Error) => {
+        reject(error);
+      };
+      strategy.authenticate(input.req, { ...options });
+    });
+  },
+
+  async authenticate(input, implementation) {
+    const { strategy, client } = await implementation;
+    const { req } = input;
+    const { searchParams } = new URL(req.url, 'https://pinniped.com');
+    const stateParam = searchParams.get('state');
+    const audience = stateParam
+      ? decodeOAuthState(stateParam).audience
+      : undefined;
+
+    return new Promise((resolve, reject) => {
+      strategy.success = user => {
+        (audience
+          ? client
+              .grant({
+                grant_type: 'urn:ietf:params:oauth:grant-type:token-exchange',
+                subject_token: user.tokenset.access_token,
+                audience,
+                subject_token_type:
+                  'urn:ietf:params:oauth:token-type:access_token',
+                requested_token_type: 'urn:ietf:params:oauth:token-type:jwt',
+              })
+              .then(tokenset => tokenset.access_token)
+              .catch(err =>
+                reject(
+                  new Error(
+                    `Failed to get cluster specific ID token for "${audience}", RFC8693 token exchange failed with error: ${err}`,
+                  ),
+                ),
+              )
+          : Promise.resolve(user.tokenset.id_token)
+        ).then(idToken => {
+          resolve({
+            fullProfile: { provider: ' ', id: ' ', displayName: ' ' },
+            session: {
+              accessToken: user.tokenset.access_token!,
+              tokenType: user.tokenset.token_type ?? 'bearer',
+              scope: user.tokenset.scope!,
+              idToken,
+              refreshToken: user.tokenset.refresh_token,
+            },
+          });
+        });
+      };
+
+      strategy.fail = info => {
+        reject(new Error(`Authentication rejected, ${info.message || ''}`));
+      };
+
+      strategy.error = (error: Error) => {
+        reject(error);
+      };
+
+      strategy.redirect = () => {
+        reject(new Error('Unexpected redirect'));
+      };
+
+      strategy.authenticate(req);
+    });
+  },
+
+  async refresh(input, implementation) {
+    const { client } = await implementation;
+    const tokenset = await client.refresh(input.refreshToken);
+
+    return new Promise((resolve, reject) => {
+      if (!tokenset.access_token) {
+        reject(new Error('Refresh Failed'));
+      }
+
+      resolve({
+        fullProfile: { provider: ' ', id: ' ', displayName: ' ' },
```

Comment:
same question re empty strings
```suggestion
        fullProfile: { provider: '', id: '', displayName: '' },
```

Reply:
Done.